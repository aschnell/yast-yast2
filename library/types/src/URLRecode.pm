#!/usr/bin/perl -w
#
# Author: Ladislav Slez√°k <lslezak@novell.com>
#
# $Id:$
#
# URLRecode.pm
#
# This is a replacement for URI::Encode perl module which cannot be used in inst-sys
# and to decrease the package dependencies
#

package URLRecode;

use strict;
use YaST::YCP qw(:LOGGING Boolean sformat);;

our %TYPEINFO;
use strict;

# local cache for char -> hex string conversion
our %escape_cache;

# fill the cache
sub InitCache
{
    for(0..255) {$escape_cache{chr($_)} = sprintf("%%%02x", $_);}
}


# Escape password, user name and fragment part of URL string
# @param input input string
# @return string Escaped string
BEGIN{ $TYPEINFO{EscapePassword} = ["function", "string", "string"];}
sub EscapePassword
{
    my ($self, $escaped) = @_;

    if (!defined %escape_cache) { InitCache(); }

    $escaped =~ s/([^A-Za-z0-9\\-_.!~*'()])/$escape_cache{$1}/ge;
    return $escaped;
}

# Escape path part of URL string
# @param input input string
# @return string Escaped string
BEGIN{ $TYPEINFO{EscapePath} = ["function", "string", "string"];}
sub EscapePath
{
    my ($self, $escaped) = @_;

    if (!defined %escape_cache) { InitCache() };

    $escaped =~ s/([^A-Za-z0-9\-_.!~*'()\/])/$escape_cache{$1}/ge;
    return $escaped;
}

# Escape query part of URL string
# @param input input string
# @return string Escaped string
BEGIN{ $TYPEINFO{EscapeQuery} = ["function", "string", "string"];}
sub EscapeQuery
{
    my ($self, $escaped) = @_;

    if (!defined %escape_cache) { InitCache(); }

    $escaped =~ s/([^A-Za-z0-9\\-_.!~*'()\/:=&])/$escape_cache{$1}/ge;
    return $escaped;
}


# UnEscape an URL string, replace %<Hexnum><HexNum> sequences
# by character
# @param input input string
# @return string Unescaped string
BEGIN{ $TYPEINFO{UnEscape} = ["function", "string", "string"];}
sub UnEscape
{
    my ($self, $input) = @_;
   
    $input =~ s/%([0-9A-Fa-f]{2})/chr(hex($1))/ge;
    return $input
}

1;
