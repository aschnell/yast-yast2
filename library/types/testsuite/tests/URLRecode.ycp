/**
 * Testsuite for URLRecode.pm module
 *
 * $Id$
 */

{

include "testsuite.ycp";
import "URLRecode";


TEST(``(URLRecode::EscapePath(nil)), [], nil);
TEST(``(URLRecode::EscapePath("")), [], nil);

string test_path = "/@#$%^&/dir/čýěščýá/file";

TEST(``(URLRecode::EscapePath(test_path)), [], nil);
TEST(``(URLRecode::UnEscape(URLRecode::EscapePath(test_path)) == test_path), [], nil);


TEST(``(URLRecode::EscapePath(" !@#$%^&*()/?+=")), [], nil);
TEST(``(URLRecode::EscapePassword(" !@#$%^&*()/?+=")), [], nil);
TEST(``(URLRecode::EscapeQuery(" !@#$%^&*()/?+=")), [], nil);

/* EOF */
}
