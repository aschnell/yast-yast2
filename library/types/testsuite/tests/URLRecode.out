Return	nil
Return	
Return	/%40%23%24%25%5e%26/dir/%c4%8d%c3%bd%c4%9b%c5%a1%c4%8d%c3%bd%c3%a1/file
Return	true
Return	%20!%40%23%24%25%5e%26*()/%3f%2b%3d
Return	%20!%40%23%24%25^%26*()%2f%3f%2b%3d
Return	%20!%40%23%24%25^&*()/%3f%2b=
